autoscale-groups-core/src/main/java/fish/payara/extensions/autoscale/groups/core/admin/ScaleCommandHelper.java

```diff
@@ -0,0 +1,327 @@
+        targetServers = clusterCommandHelper.optimizeServerListOrder(targetServers);
+
+        // Holds responses from the threads running the command
+        ArrayBlockingQueue<CommandRunnable> responseQueue = new ArrayBlockingQueue<>(nInstances);
+
+        // Make the thread pool use the smaller of the number of instances or half the admin thread pool size
+        int threadPoolSize = Math.min(nInstances, getAdminThreadPoolSize() / 2);
+        if (threadPoolSize < 1) {
+            threadPoolSize = 1;
+        }
+
+        ExecutorService threadPool = Executors.newFixedThreadPool(threadPoolSize);
+
+        if (parameterMap == null) {
+            parameterMap = new ParameterMap();
+        }
+
+        LOGGER.info(String.format(
+                "Executing %s on %d instances using a thread pool of size %d: %s", commandName, nInstances,
+                threadPoolSize, targetServers.stream().map(Server::getName).collect(Collectors.joining(", "))));
+
+        progressStatus.setTotalStepCount(nInstances);
+        progressStatus.progress(Strings.get("cluster.command.executing", commandName, nInstances));
+
+        // Loop through instance names, construct the command for each instance name, and hand it off to the threadpool.
+        for (Server server : targetServers) {
+            String serverName = server.getName();
+            waitingForServerNames.add(serverName);
+
+            ParameterMap instanceParameterMap = new ParameterMap(parameterMap);
+            // Set the instance name as the operand for the commnd
+            instanceParameterMap.set("DEFAULT", serverName);
+
+            ActionReport instanceReport = actionReport.addSubActionsReport();
+            instanceReport.setActionExitCode(SUCCESS);
+            CommandRunner.CommandInvocation invocation = commandRunner.getCommandInvocation(commandName, instanceReport,
+                    subject);
+            invocation.parameters(instanceParameterMap);
+
+            String msg = commandName + " " + serverName;
+            LOGGER.info(msg);
+
+            // Wrap the command invocation in a runnable and hand it off
+            // to the thread pool
+            CommandRunnable cmdRunnable = new CommandRunnable(invocation, instanceReport, responseQueue);
+            cmdRunnable.setName(serverName);
+            threadPool.execute(cmdRunnable);
+        }
+
+        if (LOGGER.isLoggable(FINE)) {
+            LOGGER.fine(String.format("%s commands queued, waiting for responses", commandName));
+        }
+
+        // Make sure we don't wait longer than the admin read timeout. Set our limit to be 3 seconds less.
+        long adminTimeout = RemoteRestAdminCommand.getReadTimeout() - 3000;
+        if (adminTimeout <= 0) {
+            // This should never be the case
+            adminTimeout = 57 * 1000;
+        }
+
+        if (LOGGER.isLoggable(FINE)) {
+            LOGGER.fine(String.format("Initial cluster command timeout: %d ms", adminTimeout));
```

Comment:
The message wrongly refers to clusters instead of scale groups

Reply:
Deployment groups uses the original method, so also calls them "cluster commands".
I left it as is because you're not executing a command against the scale group, you're executing it against the deployment group. It's essentially a "multi-command".

Reply:
In that case instead of "Initial cluster command..." it should say "Initial deployment group command..." instead